src/cargo/core/resolver/mod.rs

```diff
@@ -774,18 +757,50 @@ impl RemainingCandidates {
             // semver-compatible versions of a crate. For example we can't
             // simultaneously activate `foo 1.0.2` and `foo 1.2.0`. We can,
             // however, activate `1.0.2` and `2.0.0`.
-            //
-            // Here we throw out our candidate if it's *compatible*, yet not
-            // equal, to all previously activated versions.
             if let Some((a, _)) = cx.activations.get(&b_id.as_activations_key()) {
```

Comment:
I brought this up in prose, but I want to market is a concern. What ensures that a relevant package is not added to activations between `new` and `next`?

Reply:
There is nothing preventing it, but it seemingly does not matter.  

It seems to me the only reason the old implementation checks this as it iterates is to save unneeded comparisons in case of early return.

Reply:
There is nothing preventing it, but it seemingly does not matter.  

The old implementation would take a `&mut ConflictMap` in the `next()` method, but it would ONLY append new conflicts. It would not perform checks against or remove elements. As a result, it does not matter if we move the checks to `new()` and do them preemptively.

Reply:
I was not worried about `ConflictMap` getting corrupted, although maybe I should have been. If an item is missing from `ConflictMap` that did in fact prevent a candidate version from being tried that would cause big problems. Similarly, if we backtrack and construct a new `ConflictMap` it needs to know about all of the conflicts that were identified during the call to `new`.

The situation I was worried about is `next` returning a candidate that conflicts with a candidate in `activations`?
Something like:
- `new` is called for `<=1.2`. Because there is no version in `activations` `self.prioritized_candidates` is `[1.2.0, 1.1.1, 1.1.0, 1.0.0, 0.9.0 .... ]`
- Any number of things happen including discovering new dependencies and backtracking and whatever.
- `new` is called for `>=1, <2`. Because there is no version in `activations` `self.prioritized_candidates` is `[1.999.999, 1.999.998, ... , 1.1.0, 1.0.0]`
- `next` is called on the second one suggesting candidate `1.999.999`
- since there are no problems with it that version is selected and added to `activations`.
- Any number of things happen including discovering new dependencies and backtracking and whatever.
- `next` is called on the first one incorrectly suggesting candidate `1.2.0` 💥